src/Xamarin.Android.Tools.Bytecode/Kotlin/KotlinFixups.cs

```diff
@@ -39,15 +62,208 @@ public static void Fixup (IList<ClassFile> classes)
+					return method;
+			}
+
+			return null;
+		}
+
+		static MethodInfo FindJavaMethod (KotlinClass kotlinClass, KotlinFunction function, ClassFile klass)
+		{
+			var possible_methods = klass.Methods.Where (method => method.GetMethodNameWithoutSuffix () == function.Name &&
+									      method.GetFilteredParameters ().Length == function.ValueParameters.Count);
+
+			foreach (var method in possible_methods) {
+				if (!TypesMatch (method.ReturnType, function.ReturnType, kotlinClass))
+					continue;
+
+				if (!ParametersMatch (kotlinClass, method, function.ValueParameters))
+					continue;
+
+				return method;
+			}
+
+			return null;
+		}
+
+		static MethodInfo FindJavaPropertyGetter (KotlinClass kotlinClass, KotlinProperty property, ClassFile klass)
+		{
+			var possible_methods = klass.Methods.Where (method => (string.Compare (method.GetMethodNameWithoutSuffix (), $"get{property.Name}", true) == 0 ||
+									       string.Compare (method.GetMethodNameWithoutSuffix (), property.Name, true) == 0) &&
+									      method.GetParameters ().Length == 0 &&
+									      TypesMatch (method.ReturnType, property.ReturnType, kotlinClass));
+
+			return possible_methods.FirstOrDefault ();
+		}
+
+		static MethodInfo FindJavaPropertySetter (KotlinClass kotlinClass, KotlinProperty property, ClassFile klass)
+		{
+			var possible_methods = klass.Methods.Where (method => string.Compare (method.GetMethodNameWithoutSuffix (), $"set{property.Name}", true) == 0 &&
+									      property.ReturnType != null &&
+									      method.GetParameters ().Length == 1 &&
+									      TypesMatch (method.GetParameters () [0].Type, property.ReturnType, kotlinClass));
+
+			return possible_methods.FirstOrDefault ();
+		}
+
+		static bool ParametersMatch (KotlinClass kotlinClass, MethodInfo method, List<KotlinValueParameter> kotlinParameters)
+		{
+			var java_parameters = method.GetFilteredParameters ();
+
+			if (java_parameters.Length == 0 && kotlinParameters.Count == 0)
+				return true;
+
+			for (var i = 0; i < java_parameters.Length; i++) {
+				var java_p = java_parameters [i];
+				var kotlin_p = kotlinParameters [i];
+
+				if (!TypesMatch (java_p.Type, kotlin_p.Type, kotlinClass))
+					return false;
+			}
+
+			return true;
+		}
+
+		static bool TypesMatch (TypeInfo javaType, KotlinType kotlinType, KotlinClass kotlinClass)
+		{
+			// Generic type
+			if (!string.IsNullOrWhiteSpace (kotlinType.TypeParameterName) && $"T{kotlinType.TypeParameterName};" == javaType.TypeSignature)
+				return true;
+
+			if (javaType.BinaryName == KotlinUtilities.ConvertKotlinTypeSignature (kotlinType, kotlinClass))
+				return true;
+
+			// Could be a generic type erasure
+			if (javaType.BinaryName == "Ljava/lang/Object;")
+				return true;
+
+			// Sometimes Kotlin keeps its native types rather than converting them to Java native types
+			// ie: "Lkotlin/UShort;" instead of "S"
+			if (javaType.BinaryName.StartsWith ("L")) {
```

Comment:
Should this also check `javaType.BinaryName.EndsWith(';')`?

Reply:
See e.g. https://github.com/xamarin/java.interop/blob/ff3081d0241b8c84ca291d00dbc94f3a4cdf0dcc/src/Java.Interop/Java.Interop/JniTypeSignature.cs#L46

Reply:
Added.